lessons/04-structured-output/README.md

```diff
@@ -0,0 +1,590 @@
+
+ **Description**: A mill powered by a water wheel to grind grain into flour.
+
+ **Resources Needed**: Water source, wood, stone, metal for gears and axles.
+ 
+ **Effort**: High - Requires significant construction and engineering skills to build the water wheel, mill structure, and grinding mechanism. Ensuring a consistent water flow is also necessary.
+
+ ### 2. Astrolabe
+ **Title**: Astrolabe
+
+ **Description**: An astronomical instrument for solving problems related to time and star positions.
+ 
+ **Resources Needed**: Brass or other metals, engraving tools, mathematical knowledge.
+ 
+ **Effort**: Medium - Involves precise metalworking and engraving. Requires knowledge of astronomy and mathematics to ensure accuracy.
+ 
+ ### 3. Printing Press
+ **Title**: Printing Press
+ 
+ **Description**: A press for reproducing texts more efficiently than hand-copying.
+ 
+ **Resources Needed**: Wood, metal for movable type, ink, paper.
+ 
+ **Effort**: High - Building the press and creating movable type requires advanced carpentry and metalworking skills. Setting up the type and printing also involves meticulous work.
+ 
+ ### 4. Mechanical Clock
+ **Title**: Mechanical Clock
+
+ **Description**: A clock using gears and weights to keep time.
+
+ **Resources Needed**: Metal for gears, springs, weights, wood for casing, precise tools.
+
+ **Effort**: Very High - Requires intricate metalworking and a deep understanding of mechanics. Building a reliable clock involves precise assembly and calibration.
+```
+
+**Leonardo:** This is great, but who is Jason?
+
+**Time beetle:** JSON, Leonardo, JSON. It's a structured format that is easy to consume by services or applications.
+
+**Leonardo:** That's what I said, Jason.
+
+**Time beetle:** _Sigh_.
+
+As you can see, the output is structured according to the specifications in the output template. This makes it easier to read and understand and it also creates repeatable output that can be easily consumed by other services or applications.
+
+Of course a service benefits from even more structured output, like JSON, XML, or CSV so let's look at how you can ask for that.
+
+**Time beetle:** But wait, there's more!
+
+**You:** There always is with you, George. _sigh_
+
+### Even more structured output with formats like JSON, XML, or CSV
+
+**Time beetle:** So far, you've seen how you can control the output structure by instructing the model in the prompt. You have even provided an output template. Let's take that idea a step further and see how it can be applied to more structured output formats like JSON and other types of formats.
+
+Modify the prompt accordingly to ask for JSON output:
+
+> **Prompt**: Generate ideas for inventions I can build given 1500th century technology, please let me know for each what resources I need and how much effort there would be to build each invention, for each idea present in json format, include properties for title, description, resources, needed, effort
```

Comment:
I know I'm getting a bit picky - welcome to my world of "perfectionism". :-)  Updated prompt with a one-shot example.

```
< Prompt: Generate ideas for inventions that can be built using 15th-century technology. For each invention, specify the required resources and estimate the effort needed to construct it. Present the information in JSON format with the following properties: title, description, resources_needed, and effort as shown next.
> 
> Example of the JSON output:
> [
>   {
>     "title": "",
>     "description": ".",
>     "resources_needed": "",
>     "effort": ""
>   }
> ]
```